extension/extesionhelper/factory.go

```diff
@@ -0,0 +1,101 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensionhelper
+
+import (
+	"context"
+
+	"github.com/spf13/viper"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+)
+
+// FactoryOption apply changes to ExporterOptions.
+type FactoryOption func(o *factory)
+
+// CreateDefaultConfig is the equivalent of component.ExtensionFactory.CreateDefaultConfig()
+type CreateDefaultConfig func() configmodels.Extension
+
+// CreateTraceExtension is the equivalent of component.ExtensionFactory.CreateTraceExtension()
+type CreateServiceExtension func(context.Context, component.ExtensionCreateParams, configmodels.Extension) (component.ServiceExtension, error)
+
+// factory is the factory for Jaeger gRPC exporter.
+type factory struct {
+	cfgType                configmodels.Type
+	customUnmarshaler      component.CustomUnmarshaler
+	createDefaultConfig    CreateDefaultConfig
+	createServiceExtension CreateServiceExtension
+}
+
+// WithCustomUnmarshaler makes the factory to also implement component.ConfigUnmarshaler.
+func WithCustomUnmarshaler(customUnmarshaler component.CustomUnmarshaler) FactoryOption {
+	return func(o *factory) {
+		o.customUnmarshaler = customUnmarshaler
+	}
+}
+
+// NewFactory returns a component.ExtensionFactory that only supports all types.
+func NewFactory(
+	cfgType configmodels.Type,
+	createDefaultConfig CreateDefaultConfig,
+	createServiceExtension CreateServiceExtension,
+	options ...FactoryOption) component.ExtensionFactory {
+	f := &factory{
+		cfgType:                cfgType,
+		createDefaultConfig:    createDefaultConfig,
+		createServiceExtension: createServiceExtension,
+	}
+	for _, opt := range options {
+		opt(f)
+	}
+	var ret component.ExtensionFactory
+	if f.customUnmarshaler != nil {
+		ret = &factoryWithUnmarshaler{f}
+	} else {
+		ret = f
+	}
+	return ret
+}
+
+// Type gets the type of the Extension config created by this factory.
+func (f *factory) Type() configmodels.Type {
+	return f.cfgType
+}
+
+// CreateDefaultConfig creates the default configuration for processor.
+func (f *factory) CreateDefaultConfig() configmodels.Extension {
+	return f.createDefaultConfig()
+}
+
+// CreateTraceExtension creates a component.TraceExtension based on this config.
+func (f *factory) CreateExtension(
+	ctx context.Context,
+	params component.ExtensionCreateParams,
+	cfg configmodels.Extension) (component.ServiceExtension, error) {
+	return f.createServiceExtension(ctx, params, cfg)
+}
+
+var _ component.ConfigUnmarshaler = (*factoryWithUnmarshaler)(nil)
+
+type factoryWithUnmarshaler struct {
+	*factory
+}
+
+// CustomUnmarshaler returns a custom unmarshaler for the configuration or nil if
```

Comment:
Is the comment about returning `nil` actually correct here?

Reply:
Done.